[#36] Start conversion - need to look at interaction of FPV and double flux

MatthewBonanni opened the pull request.

MatthewBonanni committed:
[#36] Start conversion - need to look at interaction of FPV and double flux

MatthewBonanni commented:
@TimothyEDawson not quite ready yet but could use an eye on this when you get a chance. After reviewing the literature a bit, my understanding is that we compute an effective gamma from the current pressure and internal energy (and that changing to include chemical contribution doesn't change how we compute this gamma_star if we're doing it right). Internal energy will be properly computed in primitive_to_conservative where we set state based on the pressure.

If what I've said so far is true, FPV is still a bit complicated because of the compressible temperature / pressure reconstruction. If we start from an internal energy, we compute T with the compressible scaling from the table and then pressure from the ideal gas law. But if we need to just use gamma_star in the double flux formulation, we'll need to invert this process and set the internal energy based on the pressure in an earlier primitive_to_conservative?

Could also use an eye on the geometry source terms.

TimothyEDawson committed:
Merge branch 'develop' into 36-transport-total-energy-including-chemical

TimothyEDawson committed:
[#36] Finish implementing total energy formulation.

This update makes the use of gamma vs. gamma_star more explicit, adds
e0_star, and updates the fluxes and test cases accordingly.
Additionally, the conservative_to_primitive function was generalized to
work for both FRC and FPV modes, and could thus be moved to the
FluidPhysics base class.

TimothyEDawson commented:
Almost done! Most of the example cases work as-is and the results look reasonable. I confirmed that the FPV case runs if I pull in the changes from branch 10 which fix it, though it does blow up when the temperature gets too high. I realized that I forgot to compute internal energy from e0_star when it's available (right now it uses Cantera instead), so that should be the last thing to push before it's ready for review.

I also noticed that in its current state `ThermoTable` doesn't have a way to compute the internal energy from the temperature, which isn't difficult to add and would be a major step toward running in a non-double-flux mode. I have a separate branch where I implemented that and it wasn't too difficult, though the accuracy of the internal_energy -> temperature conversion seems to be poor at low temperatures, making the unit test I added for it fail. I'll make a separate issue for this soon.

TimothyEDawson committed:
[#36] Add double-flux get_internal_energy.

TimothyEDawson commented:
Realized I need to modify the viscous fluxes as well for the total energy transport formulation, I'll be working on that for a bit.

TimothyEDawson committed:
[#36] WIP: Update viscous flux terms

Adding the diffusive species enthalpy and viscous heating flux terms
to the viscous flux calculation. Additional modifications are still
required to support FPV approaches, as FPVTable currently lacks a species
enthalpy calculation and the species energy diffusive flux needs to be
mapped onto the flamelet scalar diffusive flux. The laminar flame case
currently blows up in the downstream region, and it's not clear yet if
that's due to a mistake in the formulation or something else.

Reworked the viscous flux to perform face-averaging first instead of
averaging the derived properties from either side. Also added a
correction velocity to the species mass diffusive fluxes to maintain
mass continuity.

TimothyEDawson commented:
@MatthewBonanni @nlaing613 @jjhans I'd appreciate a review whenever you get a chance! I still need to make some changes for the viscous fluxes to support the FPV approach, but I think the rest should be correct. I get some weird results on the laminar flame case, though, and some of that may be related to the outflow boundary condition.

TimothyEDawson committed:
[#36] Add species enthalpy calculation for ThermoTable.

TimothyEDawson commented:
Two things I'm working through right now:

1. I've found that many formulations will neglect the enthalpy flux due to mass diffusion based on various justifications. For example, [Saghafian et al. (2015)](http://dx.doi.org/10.1016/j.combustflame.2014.08.007) applies a Unity Lewis Number assumption to equation 8 to balance it against the thermal diffusion term, and [Poinsot ](https://www.amazon.com/Theoretical-Numerical-Combustion-Thierry-POINSOT/dp/2746639904) notes that when using a sensible enthalpy formulation the term is sometimes set to zero because it's negligible compared to the heat release rate. Therefore I really think this should be the user's choice whether to include it or not.

2. After some testing, I'm positive the issue with the laminar flame case is due to the diffusive flux term at the boundaries. I haven't narrowed down exactly what the issue is or how best to fix it, but if I force a zero-viscous-flux condition on the boundary it runs fine. I'm pretty sure the gradients which should be zero with the extrapolate boundary condition are non-zero.

TimothyEDawson committed:
Merge branch 'develop' into 36-transport-total-energy-including-chemical

jjhans commented:
@TimothyEDawson I'm working on this review now. I get results from laminar_flame, optimization and the validation cases (still need to sanity check the results from these) but so far have not gotten the hyshot jic case to run fully. Can you verify that I'm correctly replicating your workflow? I'm starting with #36 and then locally merging in #10 (accepting incoming changes from #10 where there are conflicts), generating the FPV table as in #10, then running hyshot jic.

TimothyEDawson commented:
@jjhans Hey James, yes that's exactly what I did to test the hyshot jic case. I haven't tested it with the latest iteration of this branch, though, it's quite likely the viscous fluxes are broken for FPV cases. That's something I'm not finished updating.

I'm thinking it might be best to close out #10 to get the JIC case running, even if the energy terms are incorrect, so that we can merge it into this branch once and for all. Because I do want to make sure this branch works with FPV before merging it in to dev.

jjhans commented:
@TimothyEDawson Okay thanks. For due diligence I'll try it again and let you know if that works or I still get issues and if I'm able to track those down - the current issue does happen to be from boundary_conditions so your intuition may be right on it being related to viscous fluxes.

And that sounds good, let me know if there's anything in particular you'd like on #10 for closeout but for now I'll focus on this.

TimothyEDawson commented:
@jjhans Well I just found some more errors in the boundary conditions which trace back to misunderstandings of the numpy indexing. Those will definitely prevent the FPV case from running, so I'm fixing them in branch 10 now, and I'll work on getting the hyshot jic case to start running over there.

TimothyEDawson committed:
Merge branch 'develop' into 36-transport-total-energy-including-chemical

TimothyEDawson commented:
Merged in the changes from branch 10. With the HyShot JIC case, I'm seeing the temperature drop below zero near the fuel injection location before the crash, so I suspect there's an issue with the energy.

TimothyEDawson committed:
[#36] Fix Z-calculation error in JICModel.

TimothyEDawson commented:
Aha! Found the issue - JICFModel.source still used an old way of calculating the mixture fraction. Correcting that enables the JIC case to run stably! @jjhans @nlaing613

TimothyEDawson committed:
[#36] Update JIC case with new geometry.

Also fixed a visualization bug due to `scalar_names` not having been
updated alongside the species transport update.

TimothyEDawson commented:
I think there's just one remaining bug to fix before this is ready to merge in, the remaining work is best split off into other issues.

In the hyshot JIC case, as the fuel mass flux increases, the pressure, density, and velocity/Mach number appear to stop changing pretty early on. The mixture fraction continues to increase as I would expect. I suspect there's either still something wrong with the source terms provided by JICModel.source, or with how the properties are being updated after they are applied. I'm going to take a closer look at them to make sure they're doing what they're intended to do.

TimothyEDawson commented:
@nlaing613 @jjhans I would recommend that we go ahead and merge this in, and open a separate issue to address fixing the issues we're observing in the JICF case.

nlaing613 reviewed (approved):
Looks good to me, nice cleanup here. I'll try to stay more on top of reviewing quickly.
